Mathlib/Topology/Algebra/Category/ProfiniteGrp.lean

```diff
@@ -162,3 +162,109 @@ instance : HasForget₂ ProfiniteGrp Grp where
   }
 
 end ProfiniteGrp
+
+/-!
+# The projective limit of finite groups is profinite
+
+* `ProfiniteGrp.limit` : the concretely constructed projective limit of finite groups
+  as a subgroup of the pi-type
+
+* `ofLimit`: projective limit of finite groups is a profinite group
+
+* Verify that the constructed limit satisfies the universal property.
+-/
+
+section Profiniteoflimit
+
+/- In this section, we prove that the projective limit of finite groups is profinite-/
+
+universe w w'
+
+namespace ProfiniteGrp
+
+variable {J : Type v} [SmallCategory J] (F : J ⥤ ProfiniteGrp.{max v w'})
+
+/-- Concretely constructing the limit of topological group as a subgroup of the  pi-type. -/
+def limit : Subgroup (Π j : J, F.obj j) where
+  carrier := {x | ∀ ⦃i j : J⦄ (π : i ⟶ j), F.map π (x i) = x j}
+  mul_mem' hx hy _ _ π := by simp only [Pi.mul_apply, map_mul, hx π, hy π]
+  one_mem' := by simp only [Set.mem_setOf_eq, Pi.one_apply, map_one, implies_true]
+  inv_mem' h _ _ π := by simp only [Pi.inv_apply, map_inv, h π]
+
+/--The functor mapping a profinite group to its underlying profinite space-/
+def profiniteGrpToProfinite : ProfiniteGrp ⥤ Profinite where
+  obj G := G.toProfinite
+  map f := ⟨f, by continuity⟩
+
+instance instProfiniteGrpToProfiniteFaithful : profiniteGrpToProfinite.Faithful := {
+  map_injective := fun {_ _} _ _ h =>
+    ConcreteCategory.hom_ext_iff.mpr (congrFun (congrArg ContinuousMap.toFun h)) }
+
+instance : CompactSpace (limit F) := inferInstanceAs
+  (CompactSpace (Profinite.limitCone (F ⋙ profiniteGrpToProfinite.{max v w'})).pt)
+
+/-- Making the direct limit of `FiniteGrp` into a `ProfiniteGrp`. -/
+def ofLimit : ProfiniteGrp := .of (ProfiniteGrp.limit F)
+
+/-- Verify that the limit constructed above exist projections to the `FiniteGrps`
+that are compatible with the morphisms between them. -/
+def limitCone : Limits.Cone F where
+  pt := ofLimit F
+  π :=
+  { app := fun j => {
+      toFun := fun x => x.1 j
+      map_one' := rfl
+      map_mul' := fun x y => rfl
+      continuous_toFun := by
+        exact (continuous_apply j).comp (continuous_iff_le_induced.mpr fun U a => a)
+    }
+    naturality := by
+      intro i j f
+      simp only [Functor.const_obj_obj, Functor.comp_obj,
+        Functor.const_obj_map, Category.id_comp, Functor.comp_map]
+      congr
+      exact funext fun x => (x.2 f).symm
+  }
+
+@[simp]
+lemma limitCone_pt : (ProfiniteGrp.limitCone F).pt =
+    ProfiniteGrp.ofLimit F := rfl
+
+@[simp, nolint simpNF]
+lemma limitCone_π_app_apply  (j : J) (x : ofLimit F) :
+    ((ProfiniteGrp.limitCone F).π.app j) x = x.1 j := rfl
```

Comment:
These can be generated from a `@[simps]` attribute on `limitCone`.

Reply:
How to solve the normal form problem?

Reply:
Tag `limitCone` with `@[simps]` and write `attribute [nolint simpNF] limitCone_π_app_toFun` below.

I think this nolint is fine, but if these lemmas aren't used in your followup PRs, I would just remove them for now. We can always add a `simps` attribute later if needed. For example, `Profinite.limitCone` doesn't have `simps`.

Reply:
sorry, how to write `attribute [nolint simpNF] ` ? Although I think removing it is fine.

Reply:
I mean

```lean
@[simps]
def limitCone ... := ...

attribute [nolint simpNF] limitCone_π_app_toFun
```

But if you think it's fine, just remove these lemmas and skip the `simps` attribute as well.